test/ffmpeg-vaapi/transcode/transcoder.py

```diff
@@ -10,35 +10,135 @@
+        codec, {}).get(mode, ([], (False, False), None))
+      return plats, reqs, ffparams
 
-  def transcode_1to1(self):
-    self.decoded = get_media()._test_artifact(
-      "{case}_{width}x{height}_{mode}.{dstextension}".format(**vars(self)))
-    get_media().test_call_timeout = vars(self).get("call_timeout", 0)
-    if vars(self).get("mode", None) == 'hwhw':
-      self.output = call(
-        "ffmpeg -hwaccel vaapi -hwaccel_device /dev/dri/renderD128 -v verbose"
-        "  -hwaccel_output_format vaapi -i {source} -an -c:v {mcodec}"
-        " -vframes {frames} -y {decoded}".format(**vars(self)))
-    elif  vars(self).get("mode", None) == 'hwsw':
-      self.output = call(
-        "ffmpeg -hwaccel vaapi -hwaccel_device /dev/dri/renderD128 -v verbose"
-        "  -hwaccel_output_format vaapi -i {source}  -vf 'hwdownload,format=nv12' -an -c:v {mcodec}"
-        " -vframes {frames} -y {decoded}".format(**vars(self)))
-    elif  vars(self).get("mode", None) == 'swhw':
-      self.output = call(
-        "ffmpeg -vaapi_device /dev/dri/renderD128 -v verbose "
-        "-i {source}  -vf 'format=nv12,hwupload' -an -c:v {mcodec}"
-        " -vframes {frames} -y {decoded}".format(**vars(self)))
+  def validate_spec(self):
+    from slash.utils.pattern_matching import Matcher
+
+    assert len(self.outputs), "Invalid test case specification, outputs data empty"
+    assert self.mode in ["sw", "hw"], "Invalid test case specification as mode type not valid"
+
+    # generate platform list based on test runtime parameters
+    iplats, ireq, _ =  self.get_requirements_data("decode", self.codec, self.mode)
+    platforms = set(iplats)
+    requires = [ireq,]
+
+    for output in self.outputs:
+      codec = output["codec"]
+      mode  = output["mode"]
+      assert mode in ["sw", "hw"], "Invalid test case specification as output mode type not valid"
+      oplats, oreq, _ = self.get_requirements_data("encode", codec, mode)
+      platforms &= set(oplats)
+      requires.append(oreq)
+
+    # create matchers based on command-line filters
+    matchers = [Matcher(s) for s in slash.config.root.run.filter_strings]
+
+    # check if user supplied any platform tag on command line
+    pmatch = any(map(lambda p: any([m.matches(p) for m in matchers]), ALL_PLATFORMS))
+
+    # if user supplied a platform tag, check if this test can support it via the
+    # test param-based required platforms
+    if pmatch and not any(map(lambda p: any([m.matches(p) for m in matchers]), platforms)):
+      slash.skip_test("unsupported platform")
+
+    # check required
+    if not all([t for t,m in requires]):
+      slash.skip_test("One or more software requirements not met")
+
+  def gen_input_opts(self):
+    opts = ""
+
+    if "hw" == self.mode:
+      opts += " -hwaccel vaapi -hwaccel_device /dev/dri/renderD128 -hwaccel_output_format vaapi"
     else:
-        assert "non supported transcoding"
+      opts += " -vaapi_device /dev/dri/renderD128"
```

Comment:
Do you think we should specify the sw decoder explicitly (e.g. `-c:v h264`) like we do in the ffmpeg-qsv tests?

Reply:
by default it is taking s/w codec in this scenario, so with or without -c:v option result is same.

Reply:
let's be explicit

Reply:
explicitly added decoder codec